src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,481 @@
+
+            public bool Contains(TKey item) => _dictionary.ContainsKey(item);
+
+            bool ICollection<TKey>.Remove(TKey item) => throw new NotSupportedException(Strings.ReadOnly_Modification);
+
+            public void CopyTo(TKey[] array, int index)
+            {
+                Entry[] entries = _dictionary._entries;
+                int i = 0;
+                int count = _dictionary._count;
+                while (count > 0)
+                {
+                    Entry entry = entries[i];
+                    if (entry.next > -2)  // part of free list?
+                    {
+                        array[index++] = entry.key;
+                        count--;
+                    }
+                    i++;
+                }
+            }
+
+            public Enumerator GetEnumerator() => new Enumerator(_dictionary); // avoid boxing
+            IEnumerator<TKey> IEnumerable<TKey>.GetEnumerator() => new Enumerator(_dictionary);
+            IEnumerator IEnumerable.GetEnumerator() => new Enumerator(_dictionary);
+
+            public struct Enumerator : IEnumerator<TKey>
+            {
+                private readonly DictionarySlim<TKey, TValue> _dictionary;
+                private int _index, _count;
+                private TKey _current;
+
+                internal Enumerator(DictionarySlim<TKey, TValue> dictionary)
+                {
+                    _dictionary = dictionary;
+                    _index = 0;
+                    _count = _dictionary._count;
+                    _current = default;
+                }
+
+                public TKey Current => _current;
+
+                object IEnumerator.Current => _current;
+
+                public void Dispose() { }
+
+                public bool MoveNext()
+                {
+                    if (_count == 0)
+                    {
+                        _current = default;
+                        return false;
+                    }
+
+                    _count--;
+
+                    while (_dictionary._entries[_index].next < -1)
+                        _index++;
+
+                    _current = _dictionary._entries[_index++].key;
+                    return true;
+                }
+
+                public void Reset()
+                {
+                    _index = 0;
+                    _count = _dictionary._count;
+                }
+            }
+        }
+
+        public struct ValueEnumerable : ICollection<TValue>, IReadOnlyCollection<TValue>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+
+            internal ValueEnumerable(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+            }
+
+            public int Count => _dictionary._count;
+
+            public bool IsReadOnly => true;
+
+            void ICollection<TValue>.Add(TValue item) => throw new NotSupportedException(Strings.ReadOnly_Modification);
```

Comment:
Replace `Strings.ReadOnly_Modification` with `Strings.NotSupported_ValueCollectionSet`? Same applies to other modification methods.

Reply:
as above